server/src/main/java/org/opensearch/gateway/ShardsBatchGatewayAllocator.java

```diff
@@ -0,0 +1,703 @@
+
+            // recalc to also (lazily) clear out old nodes.
+            this.lastSeenEphemeralIds = newEphemeralIds;
+        }
+    }
+
+    private static void clearCacheForBatchPrimary(ShardsBatch batch, RoutingAllocation allocation) {
+        // We're not running below code because for removing a node from cache we need all replica's primaries
+        // to be assigned on same node. This was easy in single shard case and we're saving a call for a node
+        // if primary was already assigned for a replica. But here we don't keep track of per shard data in cache
+        // so it's not feasible to do any removal of node entry just based on single shard.
+        // ONLY run if single shard is present in the batch, to maintain backward compatibility
+        if (batch.getBatchedShards().size() == 1) {
+            List<ShardRouting> primaries = batch.getBatchedShards()
+                .stream()
+                .map(allocation.routingNodes()::activePrimary)
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+            AsyncShardFetch<? extends BaseNodeResponse> fetch = batch.getAsyncFetcher();
+            primaries.forEach(node -> fetch.clearCacheForNode(node.currentNodeId()));
+        }
+    }
+
+    private boolean hasNewNodes(DiscoveryNodes nodes) {
+        for (final DiscoveryNode node : nodes.getDataNodes().values()) {
+            if (lastSeenEphemeralIds.contains(node.getEphemeralId()) == false) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    class InternalBatchAsyncFetch<T extends BaseNodeResponse> extends AsyncShardFetch<T> {
+        InternalBatchAsyncFetch(
+            Logger logger,
+            String type,
+            Map<ShardId, ShardAttributes> map,
+            AsyncShardFetch.Lister<? extends BaseNodesResponse<T>, T> action,
+            String batchUUId
+        ) {
+            super(logger, type, map, action, batchUUId);
+        }
+
+        @Override
+        protected void reroute(String reroutingKey, String reason) {
+            logger.trace("{} scheduling reroute for {}", reroutingKey, reason);
+            assert rerouteService != null;
+            rerouteService.reroute(
+                "async_shard_batch_fetch",
+                Priority.HIGH,
+                ActionListener.wrap(
+                    r -> logger.trace("{} scheduled reroute completed for {}", reroutingKey, reason),
+                    e -> logger.debug(new ParameterizedMessage("{} scheduled reroute failed for {}", reroutingKey, reason), e)
+                )
+            );
+        }
+    }
+
+    class InternalPrimaryBatchShardAllocator extends PrimaryShardBatchAllocator {
+
+        @Override
+        @SuppressWarnings("unchecked")
+        protected AsyncShardFetch.FetchResult<TransportNodesListGatewayStartedBatchShards.NodeGatewayStartedShardsBatch> fetchData(
```

Comment:
@khushbr @vikasvb90 @amkhar Since now we are having a separate class for batch version of GA, so tried merging this PR changes here for fetchData()
https://github.com/opensearch-project/OpenSearch/pull/8865/files

Let me know if you guys have any concern here, either I will purge the above PR or I will move my changes from here into that PR